cluster/gce/coreos: refactor node.yaml

yifan-gu opened the pull request:
Keep the installation in cloud config.
Move the configuration into a separate script.

yifan-gu commented:
This also enables the fluentd pod on the nodes.

cc @dchen1107 @jonboulle @aaronlevy

roberthbailey commented:
/cc @andyzheng0831 since this is similar to the gce trusty configs

aaronlevy commented on cluster/gce/coreos/configure-node.sh:
```diff
@@ -0,0 +1,124 @@
+}
+
+function configure-kubelet() {
+    mkdir -p /etc/kubernetes/manifests
```
I don't think you need this here. This path is also being created in configure_static_pods()

yifan-gu replied:
Good catch :)

roberthbailey commented on cluster/gce/util.sh:
```diff
@@ -386,7 +387,7 @@ function create-node-template {
       $2 \
       --can-ip-forward \
-      --metadata-from-file "$3","$4" >&2; then
+      --metadata-from-file ${@: 3:$#} >&2; then
```
From my simple local testing, this expansion doesn't seem to put commas between each argument, and gives `$3 $4` instead of `$3,$4` like we had before.

roberthbailey commented on cluster/gce/coreos/configure-node.sh:
```diff
@@ -0,0 +1,124 @@
+clusters:
+- name: local
+  cluster:
+    certificate-authority-data: ${CA_CERT}
```
Thank you for making this more secure.

yifan-gu replied:
You are very welcome!

andyzheng0831 commented on cluster/gce/coreos/helper.sh:
```diff
@@ -75,38 +75,38 @@ NUM_MINIONS: $(yaml-quote ${NUM_MINIONS})
+KUBERNETES_CONFIGURE_CBR0: $(yaml-quote ${KUBERNETES_CONFIGURE_CBR0:-true})
+CA_CERT: $(yaml-quote ${CA_CERT_BASE64})
+KUBELET_CERT: $(yaml-quote ${KUBELET_CERT_BASE64:-})
+KUBELET_KEY: $(yaml-quote ${KUBELET_KEY_BASE64:-})
```
As you move this back to yaml format, why not just use the ContainerVM code and then append whatever dedicated for coreos? That will remove unnecessary code and reduce maintenance cost.

yifan-gu replied:
Good idea. Thanks!

roberthbailey commented on cluster/gce/coreos/configure-node.sh:
```diff
@@ -0,0 +1,124 @@
+spec:
+  containers:
+  - name: fluentd-cloud-logging
+    image: gcr.io/google_containers/fluentd-gcp:1.14
```
Can you either fix this (or file a bug) to make this alternatively run the elastic search fluentd pod if the user chooses elastic search in their config-default.sh file?

yifan-gu replied:
Sure. I was thinking the cluster will always use the fluentd-gcp on gce. Thanks for pointing out!

andyzheng0831 commented on cluster/gce/coreos/helper.sh:
```diff
@@ -152,6 +152,7 @@ function create-master-instance {
-    "user-data=${KUBE_ROOT}/cluster/gce/coreos/node.yaml"
+                       "kube-env=${KUBE_TEMP}/node-kube-env.yaml" \
+                       "user-data=${KUBE_ROOT}/cluster/gce/coreos/node.yaml" \
+                       "configure-node=${KUBE_ROOT}/cluster/gce/coreos/configure-node.sh"
```
Please fix the Indentation?

roberthbailey commented on cluster/gce/coreos/configure-node.sh:
```diff
@@ -0,0 +1,124 @@
+echo "Configuring kube-proxy"
+configure-kube-proxy
+
+echo "Configuring static pods"
```
If `ENABLE_NODE_LOGGING` is set to false this should be skipped.

yifan-gu replied:
will do

andyzheng0831 commented on cluster/gce/coreos/configure-node.sh:
```diff
@@ -0,0 +1,124 @@
+EOF
+}
+
+function configure-static-pods() {
```
Hardcoding the manifest file here is an arguable method. (1) Every time people make a change to the yaml file in the salt dir need to remember to change here also; (2) If there is a new addon, we have to add the manifest file here, which will make the maintenance harder (from one file to two files). I am thinking if there is a lightweight way to just directly use the file in salt dir (And that will also benefit the trusty node config.)

yifan-gu replied:
Yeah, I have thought about this. I have even thought about to use saltstack on coreos (need to run in a container), but according to what I got from @dchen1107 the saltstack will be thrown away. So for the time being I decide to not rely on saltstack configs.

Another consideration is sometimes we will need to rewrite the pod manifest differently on coreos.
for example https://github.com/kubernetes/heapster/pull/445#discussion_r36384604

I am happy to suggestions.

andyzheng0831 replied:
Going back to salt is unwise. I raised the question as it is also a problem in the trusty config in cluster/gce/trusty/. If we can use a unified solution for different os distributions, that will be nice. I once thought about one solution: just pick out needed yaml files from dir cluster/saltbase/salt, pack them and upload to gcs bucket, then node instances download, unpack, and place files in target dir. But since the original yaml does not on coreos, I guess you have to maintain a separate version for coreos in anyway.

yifan-gu replied:
Your solution seems reasonable to me. I wish we can have some global yaml templates, and evaluating them during kube-up, then upload as you said.
But for now I want to get things working quick, so just maintaining my own copies of the yamls. It's definitely not the best solution.

andyzheng0831 replied:
Agree.

Commit:
cluster/gce: Move build-kube-env to util.sh

As now all 3 distros(coreos, debian, trusty) shares the same
build-kube-env, let's move it to util.sh.

Commit:
cluster/gce/coreos: refactor node.yaml

Keep the installation in cloud config.
Move the configuration into a separate script.

yifan-gu commented:
I fixed all the comments. But I am going to investigate if I can use the templates in the saltstack dir as discussed today with @dchen1107 .
Also CI is failing, will look into it.

yifan-gu commented:
@dchen1107 
But I am not sure how such jinja files are not evaluated by salt? Maybe I misunderstood what you were saying?

https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/kube-apiserver/kube-apiserver.manifest

yifan-gu commented:
@dchen1107 Or you actually just meant for the addons, not the master components such as api-server, controller-manger? ^^

andyzheng0831 commented:
@yifan-gu may I ask your plan and timeline for merging this PR?

yifan-gu commented:
@andyzheng0831 I would like this to be merged that ASAP.  I am currently working on the master side based on this. I think I should get api server running today, but unfortunately I have to embed the pod spec in the script for now until we have a generic solution.

andyzheng0831 commented on cluster/gce/util.sh:
```diff
@@ -386,7 +387,7 @@ function create-node-template {
       $2 \
       --can-ip-forward \
-      --metadata-from-file "$3","$4" >&2; then
+      --metadata-from-file $(echo ${@: 3: $#} | tr ' ' ',') >&2; then
```
I think "${@: 3: $#}" can be simplified as "${@:3}".

yifan-gu replied:
Thanks! Fixed.

yifan-gu commented:
close in favor of #17240 #17241 and #17243
